File: T-Bank/Algorithms-and-Data-Structures-2025/1sprint-IntroductionToSorting-BinarySearch/contest/F-Сортировка_слиянием_с_приколом/solution.cpp
#include <iostream>
#include <vector>



long long mergeAndCount(std::vector<int>& nums, std::vector<int>& tempArr, int& left, int& middle, int& right) {
    long long invCount = 0;
    int l = left;
    int r = middle + 1;
    int k = left;
    while (l <= middle && r <= right) {
        if (nums[l] < nums[r]) {
            tempArr[k] = nums[l];
            ++l;
        } else {
            tempArr[k] = nums[r];
            invCount += (middle - l + 1);
            ++r;
        }
        ++k;
    }
    while (l <= middle) {
        tempArr[k] = nums[l];
        ++l;
        ++k;
    }
    while (r <= right) {
        tempArr[k] = nums[r];
        ++r;
        ++k;
    }
    for (int i = left; i < right + 1; ++i) nums[i] = tempArr[i];
    return invCount;
}



long long mergeSortAndCount(std::vector<int>& nums, std::vector<int>& tempArr, int left, int right) {
    long long invCount = 0;
    if (left < right) {
        int middle = left + (right - left) / 2;
        invCount += mergeSortAndCount(nums, tempArr, left, middle);
        invCount += mergeSortAndCount(nums, tempArr, middle + 1, right);
        invCount += mergeAndCount(nums, tempArr, left, middle, right);
    }
    return invCount;
}



void solution() {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(N + logN)
    */
    int n;
    std::scanf("%d", &n);
    std::vector<int> nums (n, 0);
    for (int i = 0; i < n; ++i) std::cin >> nums[i];
    std::vector<int> tempArr (n, 0);
    long long invCount = mergeSortAndCount(nums, tempArr, 0, n - 1);
    std::cout << invCount << "\n";
    for (int i = 0; i < n; ++i) {
        if (i) std::cout << " ";
        std::cout << nums[i];
    }
    std::cout << "\n";
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/1sprint-IntroductionToSorting-BinarySearch/contest/F-Сортировка_слиянием_с_приколом/solution.py
import sys

def merge_and_count(arr: list[int], temp_arr: list[int], left: int, mid: int, right: int) -> int:
    i: int = left
    j: int = mid + 1
    k: int = left
    inv_count: int = 0

    while i <= mid and j <= right:
        if arr[i] <= arr[j]:
            temp_arr[k] = arr[i]
            i += 1
        else:
            temp_arr[k] = arr[j]
            inv_count += (mid - i + 1)
            j += 1
        k += 1
    while i <= mid:
        temp_arr[k] = arr[i] 
        i += 1
        k += 1
    while j <= right:
        temp_arr[k] = arr[j]
        j += 1
        k += 1
    for i in range(left, right + 1):
        arr[i] = temp_arr[i]
    return inv_count

def merge_sort_and_count(arr: list[int], temp_arr: list[int], left: int, right: int) -> int:
    inv_count: int = 0
    if left < right:
        mid: int = (left + right) // 2

        inv_count += merge_sort_and_count(arr, temp_arr, left, mid)
        inv_count += merge_sort_and_count(arr, temp_arr, mid + 1, right)
        inv_count += merge_and_count(arr, temp_arr, left, mid, right)

    return inv_count

def solution() -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(N + logN)"""
    n: int = int(sys.stdin.readline().rstrip())
    nums: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    temp_arr: list[int] = [0] * n
    inv_count = merge_sort_and_count(nums, temp_arr, 0, n - 1)
    
    sys.stdout.write(f"{inv_count}\n")
    sys.stdout.write(' '.join(map(str, nums)) + '\n')

if __name__ == '__main__':
    solution()
